database/tbcd/level/level.go

```diff
@@ -241,137 +266,372 @@ func keyToHeightHash(key []byte) (uint64, []byte) {
+		log.Infof("best hash       : %v", b)
+		log.Infof("best height     : %v", bestBH.Height)
+		log.Infof("--- FORK ---")
+	}
+
+	// Insert missing blocks and block headers
 	hhBatch := new(leveldb.Batch)
 	bmBatch := new(leveldb.Batch)
 	bhsBatch := new(leveldb.Batch)
-	for k := range bhs {
-		hhKey := heightHashToKey(bhs[k].Height, bhs[k].Hash[:])
-		// Height 0 is genesis, we do not want a missing block record for that.
-		if bhs[k].Height != 0 {
-			// Insert a synthesized height_hash key that serves as
-			// an index to see which blocks are missing.
-			bmBatch.Put(hhKey, []byte{})
+
+	cdiff := &pbh.Difficulty
+	height := pbh.Height
+	for k, bh := range bhs {
+		// The first element is skipped, as it is pre-decoded.
+		if k != 0 {
+			wbh, err = b2h(bh[:])
+			if err != nil {
+				return tbcd.ITInvalid, nil, nil,
+					fmt.Errorf("block headers insert b2h: %w", err)
+			}
+			bhash = wbh.BlockHash()
 		}
 
+		// pre set values because we start with previous value
+		height++
+		cdiff = new(big.Int).Add(cdiff, blockchain.CalcWork(wbh.Bits))
+
 		// Store height_hash for future reference
-		hhBatch.Put(hhKey, []byte{})
+		hhKey := heightHashToKey(height, bhash[:])
+		hhBatch.Put(hhKey, []byte{}) // XXX nil?
+
+		// Insert a synthesized height_hash key that serves as an index
+		// to see which blocks are missing.
+		// XXX should we always insert or should we verify prior to insert?
+		bmBatch.Put(hhKey, []byte{})
 
 		// XXX reason about pre encoding. Due to the caller code being
 		// heavily reentrant the odds are not good that encoding would
 		// only happens once. The downside is that this encoding
 		// happens in the database transaction and is thus locked.
 
-		// Encode block header as [hash][height,header] or [32][8+80] bytes
-		ebh := encodeBlockHeader(&bhs[k])
-		bhsBatch.Put(bhs[k].Hash, ebh[:])
+		// Encode block header as [hash][height,header,cdiff] or,
+		// [32][8+80+32] bytes
+		ebh := encodeBlockHeader(height, bh, cdiff)
+		bhsBatch.Put(bhash[:], ebh[:])
 		lastRecord = ebh[:]
 	}
 
-	// Insert last height into block headers XXX this does not deal with forks
-	bhsBatch.Put([]byte(bhsLastKey), lastRecord)
+	// Reason about needing to check fork flag. For now keep it here to
+	// distinguish between certain fork and maybe fork paths.
+	var (
+		it       tbcd.InsertType
+		cbh, lbh *tbcd.BlockHeader
+	)
+	if fork {
+		// Insert last height into block headers if the new cumulative
+		// difficulty exceeds the prior difficulty.
+		switch cdiff.Cmp(&bestBH.Difficulty) {
+		case -1, 0:
+			// Extend fork, fork did not overcome difficulty
+			it = tbcd.ITForkExtend
+
+			// XXX should we return old best block header here?
+			// That way the caller can do best vs previous best diff.
+			log.Debugf("(%v) : %v <= %v", height, cdiff, bestBH.Difficulty)
+			cbh = bestBH
+			var header [80]byte
+			lbh = &tbcd.BlockHeader{
+				Hash:       bhash[:],
+				Height:     height,
+				Header:     header[:],
+				Difficulty: *cdiff,
+			}
```

Comment:
Could this be de-duplicated by initialising `cbh` to this value, and initialising `lbh = cbh`
Then setting `cbh = bestBH` in this case?

Reply:
Yes, now that we know that it works that needs to be cleaned up.